libraries/lib-time-and-pitch/FormantShifter.cpp

```diff
@@ -0,0 +1,183 @@
+#include "FormantShifter.h"
+#include "FormantShifterLoggerInterface.h"
+#include <algorithm>
+#include <fstream>
+
+namespace
+{
+constexpr auto MapToPositiveHalfIndex(int index, int fullSize)
+{
+   const auto inputIsValid = fullSize > 0 && fullSize % 2 == 0;
+   if (!inputIsValid)
+      return 0;
+   if (index >= 0)
+      index = index % fullSize;
+   else
+      index = fullSize - (-index % fullSize);
+   if (index > fullSize / 2)
+      index = fullSize - index;
+   return index;
+}
+
+static_assert(MapToPositiveHalfIndex(-7, 4) == 1);
+static_assert(MapToPositiveHalfIndex(9, 4) == 1);
+
+constexpr float GetLog2(float x)
+{
+   // https://stackoverflow.com/a/28730362
+   // Precise up to two decimal places. If given unnormalized power, this
+   // shouldn't be a problem.
+   union
+   {
+      float val;
+      int32_t x;
+   } u = { x };
+   auto log_2 = (float)(((u.x >> 23) & 255) - 128);
+   u.x &= ~(255 << 23);
+   u.x += 127 << 23;
+   log_2 += ((-0.3358287811f) * u.val + 2.0f) * u.val - 0.65871759316667f;
+   return log_2;
+}
+
+// `x` has length `fftSize/2+1`.
+// Returns the last bin that wasn't zeroed.
+size_t ResampleFreqDomain(float* x, size_t fftSize, double factor)
+{
+   const auto size = fftSize / 2 + 1;
+   const auto end = std::min(size, size_t(size * factor));
+   std::vector<float> tmp(end);
+   for (size_t i = 0; i < end; ++i)
+   {
+      const int int_pos = i / factor;
+      const float frac_pos = 1.f * i / factor - int_pos;
+      const auto k = MapToPositiveHalfIndex(int_pos, fftSize);
+      const auto l = MapToPositiveHalfIndex(int_pos + 1, fftSize);
+      tmp[i] = (1 - frac_pos) * x[k] + frac_pos * x[l];
+   }
+   std::copy(tmp.begin(), tmp.begin() + end, x);
+   if (end < size)
+      std::fill(x + end, x + size, 0.f);
+   return end;
+}
+} // namespace
+
+FormantShifter::FormantShifter(
+   int sampleRate, double cutoffQuefrency,
+   FormantShifterLoggerInterface& logger)
+    : cutoffQuefrency { cutoffQuefrency }
+    , mSampleRate { sampleRate }
+    , mLogger { logger }
+{
+}
+
+void FormantShifter::Reset(size_t fftSize)
+{
+   mFft = std::make_unique<staffpad::audio::FourierTransform>(fftSize);
+   const auto numBins = fftSize / 2 + 1;
+   mEnvelope.setSize(1, numBins);
+   mCepstrum.setSize(1, fftSize);
+   mEnvelopeReal.resize(numBins);
+   mWeights.resize(numBins);
+}
+
+void FormantShifter::Reset()
+{
+   mFft.reset();
+}
+
+void FormantShifter::Process(
+   const float* powSpec, std::complex<float>* spec, double factor)
+{
+   if (cutoffQuefrency == 0 || !mFft)
+      return;
+
+   const auto fftSize = mFft->getSize();
+   const auto numBins = fftSize / 2 + 1;
+
+   mLogger.Log(fftSize, "fftSize");
+
+   // Take the log of the normalized magnitude. (This assumes that
+   // the window averages to 1.)
+   std::complex<float>* pEnv = mEnvelope.getPtr(0);
+   const float normalizer = GetLog2(fftSize);
+   std::transform(powSpec, powSpec + numBins, pEnv, [&](float power) {
+      // Why -12 ?
```

Comment:
Why?

Reply:
A remnant of me copying chunks from the [stftPitchShift repo](https://github.com/jurihock/stftPitchShift/blob/48042d72a43287676543950691d227732a46759b/cpp/StftPitchShift/Cepster.h#L48) to get started.
It's not a great protection, because, as you commented someplace else, the weights can still end up being huge. I expect this can happen if for some reason the input signal has a frequency region of near-zero values.
I'm about to push a protection that I think makes sense: the purpose of formant preservation is to boost or attenuate spectral regions by some reasonable amount of decibels. This commit will limit the amplification to +20dB, which is plenty for our purpose and then avoid crazy amplification.